File: ios/EcoBuilding/Sources/ReportButton.swift
import QuickLook
import SwiftUI
import UIKit

/// Obtenir la fiche PDF — **l'élément le plus visible de l'écran**.
///
/// C'est l'objet que l'utilisateur emporte : celui qu'il envoie à un notaire, à
/// un artisan, ou qu'il garde avant une visite. Tout le reste de la fiche sert à
/// donner envie de l'obtenir ; ce bouton ne doit donc jamais être une ligne
/// discrète en bas d'écran.
///
/// La génération prend 10 à 45 s côté serveur (données, rendu 3D, mise en page).
/// On annonce les étapes réelles plutôt qu'une barre de progression lisse, qui
/// serait une fiction : un rendu serveur unique n'expose aucune progression.
struct ReportButton: View {
    @ObservedObject var model: BuildingModel
    var autoStart: Binding<Bool> = .constant(false)
    @State private var phase: Phase = .idle
    @State private var elapsed = 0
    @State private var pdf: URL?
    @State private var timer: Timer?
    @State private var quota: API.Quota?
    /// Angle de la flèche pendant la préparation. Une icône figée dix à
    /// quarante-cinq secondes laisse croire que rien ne se passe — le
    /// mouvement est ce qui distingue « ça travaille » de « c'est bloqué ».
    @State private var spinning = false

    enum Phase: Equatable {
        case idle, running, failed
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Button(action: download) {
                HStack(spacing: 10) {
                    Image(systemName: phase == .running
                          ? "arrow.triangle.2.circlepath" : "doc.text.fill")
                        .rotationEffect(.degrees(spinning ? 360 : 0))
                        .animation(spinning
                                   ? .linear(duration: 1).repeatForever(autoreverses: false)
                                   : .default, value: spinning)
                    Text(phase == .running ? stageLabel : t("get_report"))
                        .fontWeight(.semibold)
                    Spacer()
                    if phase == .running {
                        Text("\(elapsed) s").font(.footnote).opacity(0.8)
                    }
                }
                .padding(.vertical, 14).padding(.horizontal, 16)
                .frame(maxWidth: .infinity)
                .background(Color(red: 0.17, green: 0.48, blue: 0.29),
                            in: RoundedRectangle(cornerRadius: 12))
                .foregroundStyle(.white)
            }
            .disabled(phase == .running || model.buildingID == nil)

            if phase == .failed {
                Text(t("report_failed"))
                    .font(.footnote).foregroundStyle(.orange)
            } else if let left = quota?.summary(for: model.buildingID) {
                // Dire ce qu'il reste AVANT d'en manquer : on découvrait la
                // limite en la heurtant. Aucun prix affiché tant que le mur
                // payant n'existe pas — annoncer un tarif qu'on ne peut pas
                // encaisser serait un mensonge.
                Text(left)
                    .font(.footnote)
                    .foregroundStyle(quota?.blocked == true ? .orange : .secondary)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
        // PLEIN ÉCRAN, et non une feuille : la bande de carte laissée visible
        // en haut donnait à croire qu'on pouvait y revenir en la touchant, ce
        // qu'une feuille système ne permet pas. Ou l'on voit la carte et on
        // peut la toucher, ou on ne la voit pas.
        .fullScreenCover(item: $pdf) { url in PDFPreview(url: url) }
        .task { quota = try? await API.quota() }
        // Double appui sur la carte : la fiche part dès que le bâtiment a
        // répondu, sans passer par le bouton.
        .onChange(of: model.buildingID) { id in
            if autoStart.wrappedValue, id != nil, phase != .running {
                autoStart.wrappedValue = false
                download()
            }
        }
    }

    private var stageLabel: String {
        // Étapes calées sur les durées réelles observées côté serveur.
        switch elapsed {
        case ..<3: return t("stage_collect")
        case 3..<12: return t("stage_render")
        default: return t("stage_layout")
        }
    }

    private func download() {
        guard let id = model.buildingID else { return }
        phase = .running
        spinning = true
        elapsed = 0
        timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { _ in
            Task { @MainActor in elapsed += 1 }
        }
        Task {
            defer { timer?.invalidate() }
            do {
                let url = try await API.report(buildingID: id, lon: model.lon, lat: model.lat)
                phase = .idle
                spinning = false
                pdf = url                       // ouvre le lecteur
                quota = try? await API.quota()  // le solde vient de changer
            } catch {
                phase = .failed
                spinning = false
            }
        }
    }
}

/// Lecteur natif d'iOS : la fiche s'AFFICHE tout de suite.
///
/// La feuille de partage seule obligeait à enregistrer dans Fichiers puis à
/// rouvrir depuis le gestionnaire de fichiers, qui la confiait à un navigateur —
/// trois étapes et une sortie de l'app pour voir ce qu'on vient de payer. Ici la
/// fiche s'ouvre directement, et le bouton de partage du lecteur permet ensuite
/// de l'envoyer à un client ou de l'enregistrer.
struct PDFPreview: View {
    let url: URL
    @Environment(\.dismiss) private var dismiss

    var body: some View {
        VStack(spacing: 0) {
            // En-tête À NOUS. La croix posée sur le QLPreviewController n'était
            // jamais rendue : celui-ci gère sa propre barre et masque celle
            // qu'on lui fournit. L'utilisateur restait coincé sur le document
            // qu'il venait d'obtenir, juste après l'action la plus importante
            // de l'app — seul le glissement vers le bas s'en sortait, et rien
            // ne l'annonçait.
            HStack {
                Button {
                    dismiss()
                } label: {
                    Image(systemName: "xmark.circle.fill")
                        .font(.title2)
                        .foregroundStyle(.secondary)
                }
                .accessibilityLabel(t("sheet_close"))
                Spacer()
                ShareLink(item: url) {
                    Image(systemName: "square.and.arrow.up").font(.title3)
                }
                .accessibilityLabel(t("report_share"))
            }
            .padding(.horizontal, 16).padding(.vertical, 10)
            .background(.regularMaterial)

            QuickLook(url: url)
        }
    }
}

/// Lecteur natif d'iOS : la fiche s'AFFICHE tout de suite.
///
/// La feuille de partage seule obligeait à enregistrer dans Fichiers puis à
/// rouvrir depuis le gestionnaire de fichiers, qui la confiait à un navigateur —
/// trois étapes et une sortie de l'app pour voir ce qu'on vient de payer.
private struct QuickLook: UIViewControllerRepresentable {
    let url: URL

    func makeUIViewController(context: Context) -> QLPreviewController {
        let preview = QLPreviewController()
        preview.dataSource = context.coordinator
        return preview
    }
    func updateUIViewController(_ vc: QLPreviewController, context: Context) {}
    func makeCoordinator() -> Coordinator { Coordinator(url: url) }

    final class Coordinator: NSObject, QLPreviewControllerDataSource {
        let url: URL
        init(url: URL) { self.url = url }
        func numberOfPreviewItems(in controller: QLPreviewController) -> Int { 1 }
        func previewController(_ controller: QLPreviewController,
                               previewItemAt index: Int) -> QLPreviewItem {
            url as QLPreviewItem
        }
    }
}

extension URL: @retroactive Identifiable {
    public var id: String { absoluteString }
}

File: android/app/src/main/java/io/confinia/ecobuilding/ReportButton.kt
package io.confinia.ecobuilding

import android.content.Context
import android.content.Intent
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Autorenew
import androidx.compose.material.icons.filled.Description
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.FileProvider
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.io.File

/**
 * Obtenir la fiche PDF — **l'élément le plus visible de l'écran**.
 *
 * C'est l'objet que l'utilisateur emporte : celui qu'il envoie à un notaire, à
 * un artisan, ou qu'il garde avant une visite. Tout le reste de la fiche sert à
 * donner envie de l'obtenir ; ce bouton ne doit donc jamais être une ligne
 * discrète en bas d'écran.
 *
 * La génération prend 10 à 45 s côté serveur (données, rendu 3D, mise en page).
 * On annonce les étapes réelles plutôt qu'une barre de progression lisse, qui
 * serait une fiction : un rendu serveur unique n'expose aucune progression.
 */
@Composable
fun ReportButton(model: BuildingModel, quota: Quota?, onQuotaChanged: (Quota?) -> Unit) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    var running by remember { mutableStateOf(false) }
    var elapsed by remember { mutableIntStateOf(0) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(running) {
        elapsed = 0
        while (running && isActive) { delay(1000); elapsed += 1 }
    }

    Column(Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Button(
            onClick = {
                val id = model.buildingId ?: return@Button
                error = null
                running = true
                scope.launch {
                    try {
                        val file = Api.report(context, id, model.lon, model.lat)
                        openPdf(context, file)
                        onQuotaChanged(runCatching { Api.quota(context) }.getOrNull())
                    } catch (e: ReportError) {
                        // Le serveur sait pourquoi il refuse (limite atteinte,
                        // paiement requis) : son message vaut mieux que le nôtre.
                        error = e.detail.ifBlank { "La fiche n'a pas pu être générée. Réessayez." }
                    } catch (e: Exception) {
                        error = "La fiche n'a pas pu être générée. Réessayez."
                    } finally {
                        running = false
                    }
                }
            },
            enabled = !running && model.buildingId != null,
            modifier = Modifier.fillMaxWidth().height(52.dp),
            // La couleur du TEXTE doit être dite : sans elle, Material peint le
            // libellé dans la teinte du thème — du violet sur du vert, illisible.
            colors = ButtonDefaults.buttonColors(
                containerColor = Color(0.17f, 0.48f, 0.29f), contentColor = Color.White),
        ) {
            Icon(if (running) Icons.Filled.Autorenew else Icons.Filled.Description,
                contentDescription = null)
            Spacer(Modifier.width(10.dp))
            Text(if (running) stageLabel(elapsed) else "Obtenir la fiche PDF",
                fontWeight = FontWeight.SemiBold)
            if (running) {
                Spacer(Modifier.weight(1f))
                Text("$elapsed s", fontSize = 12.sp)
            }
        }

        val message = error ?: quotaLine(quota, model.buildingId)
        if (message != null) {
            // Dire ce qu'il reste AVANT d'en manquer : on découvrait la limite
            // en la heurtant. Aucun prix affiché tant que le mur payant n'existe
            // pas — annoncer un tarif qu'on ne peut pas encaisser serait faux.
            Text(message, fontSize = 12.sp,
                color = if (error != null || quota?.reportsLeft == 0)
                    Color(0.9f, 0.5f, 0.1f) else Color.Gray)
        }
    }
}

/** Étapes calées sur les durées réelles observées côté serveur. */
private fun stageLabel(elapsed: Int): String = when {
    elapsed < 3 -> "Collecte des données…"
    elapsed < 12 -> "Rendu de la carte 3D…"
    else -> "Mise en page…"
}

/**
 * Ouvre la fiche dans le lecteur PDF du téléphone.
 *
 * Le fichier passe par un FileProvider : depuis Android 7, transmettre un
 * `file://` à une autre application lève une exception. Et sans lecteur PDF
 * installé — c'est le cas d'origine sur beaucoup d'appareils — on propose le
 * partage, qui permet au moins d'envoyer ou d'enregistrer la fiche plutôt que
 * de laisser l'utilisateur devant un échec muet.
 */
private fun openPdf(context: Context, file: File) {
    val uri = FileProvider.getUriForFile(context, "${context.packageName}.files", file)
    val view = Intent(Intent.ACTION_VIEW)
        .setDataAndType(uri, "application/pdf")
        .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_ACTIVITY_NEW_TASK)
    if (view.resolveActivity(context.packageManager) != null) {
        context.startActivity(view)
        return
    }
    val share = Intent(Intent.ACTION_SEND)
        .setType("application/pdf")
        .putExtra(Intent.EXTRA_STREAM, uri)
        .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
    context.startActivity(Intent.createChooser(share, "Ouvrir la fiche PDF")
        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
}
